### PR TITLE
mirny: hw_rev independent, human readable clk_sel

### LINKED ISSUE #1559: Mirny: clock selection documentation
# Bug Report

## One-Line Summary

Documentations in `mirny.py` about clock selection differs from what is mentioned in the schematics of Mirny v1.1.

## Issue Details
The comments in question:
```
    :param clk_sel: Reference clock selection.
        valid options are: 0 - internal 100MHz XO; 1 - front-panel SMA; 2 -
        internal MMCX
```
According to the [schematics for Mirny v1.1](https://github.com/sinara-hw/Mirny/releases), both 1 and 3 should refer internal MMCX, 2 should corresponds to the front-panel SMA.

Follow-up: #1530 @airwoodix

### PR DESCRIPTION
**Pull request by airwoodix**
# ARTIQ Pull Request

## Description of Changes

This patch allows for hardware-revision independent user strings instead of integers for the Mirny reference clock selection. The relevant integer is calculated from the hardware revision read from the CPLD at ``init()``. The change is fully backwards compatible. The ``clk_sel`` functionality is identical to pre-existing code, but the interface is nicer.

Tested against v1.0 and v1.1 hardware on all three global clock sources.

``hw_rev`` is not tested anymore in ``init()``. Removed corresponding fix in ``artiq_sinara_tester``.

### Related Issue

Closes #1559 (updated docstring)

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |
| ✓  | :sparkles: New feature |
|  | :hammer: Refactoring  |
|  | :scroll: Docs |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Update [RELEASE_NOTES.rst](../RELEASE_NOTES.rst) if there are noteworthy changes, especially if there are changes to existing APIs.
- [x] Close/update issues.
- [x] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [x] Add and check docstrings and comments

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

File: artiq/coredevice/mirny.py [hunk 1]
```diff
@@ -37,22 +40,39 @@ class Mirny:
...
         self.bus = dmgr.get(spi_device)
 
+        # reference clock frequency
         self.refclk = refclk
         assert 10 <= self.refclk / 1e6 <= 600, "Invalid refclk"
 
-        self.clk_sel = clk_sel & 0b11
-        assert 0 <= self.clk_sel <= 3, "Invalid clk_sel"
+        # reference clock selection
+        if isinstance(clk_sel, str):
```
Comment (jordens) [inline artiq/coredevice/mirny.py:62]: Duck typing. `try: ... clk_sel.lower()... except TypeError: ...`?

File: artiq/coredevice/mirny.py [hunk 2]
```diff
@@ -78,15 +98,16 @@ def init(self, blind=False):
 
         :param blind: Do not attempt to verify presence and compatibility.
```
Comment (jordens) [inline artiq/coredevice/mirny.py:99]: The effect of `blind` on the clk_sel choice should be documented.

File: artiq/coredevice/mirny.py [hunk 3]
```diff
@@ -78,15 +98,16 @@ def init(self, blind=False):
...
-            reg0 = self.read_reg(0)
-            if reg0 & 0b11 != 0b11:
-                raise ValueError("Mirny HW_REV mismatch")
-            if (reg0 >> 2) & 0b11 != 0b00:
+            if (reg0 >> 2) & 0x3 != PROTO_REV_MATCH:
                 raise ValueError("Mirny PROTO_REV mismatch")
             delay(100 * us)  # slack
 
         # select clock source
+        self.clk_sel = self.clk_sel_hw_rev[self.hw_rev - 2]
```
Comment (jordens) [inline artiq/coredevice/mirny.py:110]: With `hw_rev` in `[0, 1, 2, 3]`, this is `[-2, -1, 0, 1]`, i.e. `[0, 1, 0, 1]` silently wrapping around. Does this make sense? I'd rather have `clk_sel_hw_rev` be a full 4-element array (with error values where needed).

File: artiq/frontend/artiq_sinara_tester.py [hunk 4]
```diff
@@ -291,18 +291,7 @@ def test_urukuls(self):
...
-            raise ValueError("Mirny HW_REV mismatch")
-        if (reg0 >> 2) & 0b11 != 0b00:
-            raise ValueError("Mirny PROTO_REV mismatch")
-        delay(100 * us)  # slack
-
-        # select clock source
-        cpld.write_reg(1, (cpld.clk_sel << 4))
-        delay(1000 * us)
-        # End of modified Mirny.init()
+        cpld.init()
```
Comment (jordens) [inline artiq/frontend/artiq_sinara_tester.py:294]: Good!

File: artiq/coredevice/mirny.py (hunk 2 above)
Comment (airwoodix) [reply artiq/coredevice/mirny.py]: Is there an effect of ``blind`` on the ``clk_sel`` choice? There shouldn't.

File: artiq/coredevice/mirny.py (hunk 3 above)
Comment (airwoodix) [reply artiq/coredevice/mirny.py]: Does d691b05 address this appropriately?

**Comment by airwoodix** [comment]
Thanks for the feedback. The latest commit adds better overall error handling for the ``clk_sel`` selection (and consistency: ``ValueError`` is raised in all cases except when the ``__init__`` parameter is neither a string nor an integer, in which case it's ``TypeError`` from the bitwise-and).

File: artiq/coredevice/mirny.py (hunk 2 above)
Comment (jordens) [reply artiq/coredevice/mirny.py]: Ah. I see that you have taken the read out of the blind branch. All good.

File: artiq/coredevice/mirny.py (hunk 3 above)
Comment (jordens) [reply artiq/coredevice/mirny.py]: Yes. Thanks.

**Comment by jordens** [comment]
Thanks for the PR!

**Pull request merged**